lib/martian_robot.rb

```diff
@@ -13,24 +13,50 @@ def initialize(coordinates, orientation, planet)
   end
 
   def report_final_position(instruction)
-    final_position = starting_position
-    instruction.split('').each {|single_instruction| execute(single_instruction, final_position)}
-    lost?(final_position) ? :lost : final_position
+    instruction.split('').each {|single_instruction|
+      execute(single_instruction)
+      break if lost_before_executing_all_instruction(@coordinates)
+    }
+    if lost?(@coordinates)
+      stop_or_get_lost
+    else
+      @coordinates
+    end
   end
 
-
-  def starting_position
-    x = @coordinates[0]
-    y = @coordinates[1]
-    [x,y]
+  def last_position_before_lost
+    last_position = [@coordinates[0], @coordinates[1]]
+    if @coordinates[0] > @planet.size
```

Comment:
I'm thinking, to make the code easier to read, rather than always saying `@coordinates[0]` and `@coordinates[1]`, you could have two methods called `x_coordinate` and `y_coordinate` inside which you do `@coordinates[0]` and `@coordinates[1]` and then always use those methods.